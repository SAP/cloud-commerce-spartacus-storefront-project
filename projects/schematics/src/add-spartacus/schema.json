{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "baseUrl": {
      "type": "string",
      "description": "The url of the OCC backend",
      "default": "https://localhost:9002"
    },
    "baseSite": {
      "type": "string",
      "description": "The base site to use with spartacus",
      "default": "electronics"
    },
    "useMetaTags": {
      "type": "boolean",
      "description": "Configure baseUrl via meta tag in index.html"
    },
    "featureLevel": {
      "type": "string",
      "description": "Configure feature level for application.",
      "default": "1.1"
    },
    "overwriteAppComponent": {
      "type": "boolean",
      "description": "Overwrite whole app.component.html file.",
      "default": false
    }
  },
  "required": []
}
