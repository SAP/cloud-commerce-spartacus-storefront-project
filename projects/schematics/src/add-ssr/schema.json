{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    }
  },
  "required": []
}
