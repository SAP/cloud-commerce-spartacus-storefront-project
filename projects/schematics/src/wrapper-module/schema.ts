import { ExecutionOptions } from '@angular-devkit/schematics';

export interface Schema extends Partial<ExecutionOptions> {
  /**
   * The name of the project in which to execute schematics.
   */
  project: string;
  /**
   * When enabled, prints the additional logs.
   */
  debug?: boolean;
  /**
   * The marker module for which
   * to find or generate a new wrapper module.
   */
  markerModuleName: string;
  /**
   * Name of the feature module
   * to add to the wrapper module.
   */
  featureModuleName: string;
  /**
   * Internal options.
   * Should not be set by the user.
   */
  internal?: {
    /**
     * The execution sequence.
     * Used for ensuring one log message is printed.
     */
    sequence?: number;
  };
}
