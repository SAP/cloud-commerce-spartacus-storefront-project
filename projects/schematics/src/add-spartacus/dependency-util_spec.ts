/// <reference types="jest" />

import {
  CLI_CART_BASE_FEATURE,
  CLI_CDC_FEATURE,
  CLI_CHECKOUT_B2B_FEATURE,
  CLI_CHECKOUT_BASE_FEATURE,
  CLI_DIGITAL_PAYMENTS_FEATURE,
  CLI_ORDER_FEATURE,
  CLI_USER_ACCOUNT_FEATURE,
  CLI_USER_PROFILE_FEATURE,
  SPARTACUS_ASM,
  SPARTACUS_CART,
  SPARTACUS_CDC,
  SPARTACUS_CHECKOUT,
  SPARTACUS_DIGITAL_PAYMENTS,
  SPARTACUS_ORDER,
  SPARTACUS_USER,
} from '../shared/libs-constants';
import {
  analyzeCrossFeatureDependencies,
  analyzeCrossLibraryDependencies,
} from './dependency-util';

describe('dependency-util', () => {
  describe('analyzeCrossFeatureDependencies', () => {
    it('DP - should return the correct set of ordered libraries', () => {
      const result = analyzeCrossFeatureDependencies([
        CLI_DIGITAL_PAYMENTS_FEATURE,
        CLI_CHECKOUT_B2B_FEATURE,
        CLI_CDC_FEATURE,
      ]);

      expect(result[SPARTACUS_DIGITAL_PAYMENTS]).toEqual([
        CLI_DIGITAL_PAYMENTS_FEATURE,
      ]);
      expect(result[SPARTACUS_CHECKOUT]).toEqual([
        CLI_CHECKOUT_BASE_FEATURE,
        CLI_CHECKOUT_B2B_FEATURE,
      ]);
      expect(result[SPARTACUS_CART]).toEqual([CLI_CART_BASE_FEATURE]);
      expect(result[SPARTACUS_USER]).toEqual([
        CLI_USER_PROFILE_FEATURE,
        CLI_USER_ACCOUNT_FEATURE,
      ]);
      expect(result[SPARTACUS_ORDER]).toEqual([CLI_ORDER_FEATURE]);
      expect(result[SPARTACUS_CDC]).toEqual([CLI_CDC_FEATURE]);
    });
    it('User Profile - should return the correct set of ordered libraries', () => {
      const result = analyzeCrossFeatureDependencies([
        CLI_USER_PROFILE_FEATURE,
      ]);

      expect(result[SPARTACUS_USER]).toEqual([
        CLI_USER_PROFILE_FEATURE,
        CLI_USER_ACCOUNT_FEATURE,
      ]);
    });
  });

  describe('analyzeCrossLibraryDependencies', () => {
    it('DP - should return the correct set of ordered libraries', () => {
      const result = analyzeCrossLibraryDependencies([
        SPARTACUS_DIGITAL_PAYMENTS,
      ]);

      expect(result).toEqual([
        SPARTACUS_USER,
        SPARTACUS_CART,
        SPARTACUS_ORDER,
        SPARTACUS_CHECKOUT,
        SPARTACUS_DIGITAL_PAYMENTS,
      ]);
    });
    it('CDC - should return the correct set of ordered libraries', () => {
      const result = analyzeCrossLibraryDependencies([SPARTACUS_CDC]);

      expect(result).toEqual([SPARTACUS_USER, SPARTACUS_ASM, SPARTACUS_CDC]);
    });
  });
});
