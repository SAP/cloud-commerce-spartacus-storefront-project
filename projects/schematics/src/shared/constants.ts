export const UTF_8 = 'utf-8';

/***** Imports start *****/
export const ANGULAR_SCHEMATICS = '@schematics/angular';
export const ANGULAR_CORE = '@angular/core';
export const ANGULAR_FORMS = '@angular/forms';
export const ANGULAR_ROUTER = '@angular/router';
export const RXJS = 'rxjs';

export const SPARTACUS_CORE = '@spartacus/core';
export const SPARTACUS_STOREFRONTLIB = '@spartacus/storefront';
export const SPARTACUS_CDS = '@spartacus/cds';

export const NGRX_STORE = '@ngrx/store';
/***** Imports end *****/

/***** Classes start *****/
export const OBSERVABLE_CLASS = 'Observable';

export const STORE = 'Store';

export const FORM_BUILDER = 'FormBuilder';

export const CHANGE_DETECTOR_REF = 'ChangeDetectorRef';

export const ACTIVATED_ROUTE = 'ActivatedRoute';
export const ROUTER = 'Router';

export const CMS_COMPONENT_DATA_CLASS = 'CmsComponentData';
export const CONFIG_MODULE_CLASS = 'ConfigModule';
export const CMS_CONFIG = 'CmsConfig';
export const CMS_SELECTORS = 'CmsSelectors';
export const CMS_ACTIONS = 'CmsActions';
export const LOAD_CMS_COMPONENT_CLASS = 'LoadCmsComponent';
export const LOAD_CMS_COMPONENT_FAIL_CLASS = 'LoadCmsComponentFail';
export const LOAD_CMS_COMPONENT_SUCCESS_CLASS = 'LoadCmsComponentSuccess';
export const CMS_GET_COMPONENT_FROM_PAGE = 'CmsGetComponentFromPage';
export const USER_ADDRESS_SERVICE = 'UserAddressService';
export const AUTH_SERVICE = 'AuthService';
export const FEATURE_CONFIG_SERVICE = 'FeatureConfigService';
export const PAGE_META_RESOLVER = 'PageMetaResolver';
export const CMS_SERVICE = 'CmsService';
export const PAGE_META_SERVICE = 'PageMetaService';
export const CHECKOUT_SERVICE = 'CheckoutService';
export const CHECKOUT_PAYMENT_SERVICE = 'CheckoutPaymentService';
export const CHECKOUT_DELIVERY_SERVICE = 'CheckoutDeliveryService';
export const CART_DATA_SERVICE = 'CartDataService';
export const CART_SERVICE = 'CartService';
export const ACTIVE_CART_SERVICE = 'ActiveCartService';
export const PROMOTION_SERVICE = 'PromotionService';
export const ORDER_DETAILS_SERVICE = 'OrderDetailsService';
export const CHECKOUT_LOGIN_COMPONENT = 'CheckoutLoginComponent';
export const AUTH_REDIRECT_SERVICE = 'AuthRedirectService';
export const CHECKOUT_DETAILS_SERVICE = 'CheckoutDetailsService';
export const NOT_CHECKOUT_AUTH_GUARD = 'NotCheckoutAuthGuard';
export const ROUTING_SERVICE = 'RoutingService';
export const SHIPPING_ADDRESS_COMPONENT = 'ShippingAddressComponent';
export const CHECKOUT_CONFIG_SERVICE = 'CheckoutConfigService';
export const TRANSLATION_SERVICE = 'TranslationService';
export const CHECKOUT_PAGE_META_RESOLVER = 'CheckoutPageMetaResolver';
export const ADD_TO_CART_COMPONENT = 'AddToCartComponent';
export const MODAL_SERVICE = 'ModalService';
export const CURRENT_PRODUCT_SERVICE = 'CurrentProductService';
export const CART_NOT_EMPTY_GUARD = 'CartNotEmptyGuard';
export const CART_TOTALS_COMPONENT = 'CartTotalsComponent';
export const MINI_CART_COMPONENT = 'MiniCartComponent';
export const CHECKOUT_ORDER_SUMMARY_COMPONENT = 'CheckoutOrderSummaryComponent';
export const CHECKOUT_PROGRESS_MOBILE_TOP_COMPONENT =
  'CheckoutProgressMobileTopComponent';
export const CHECKOUT_CONFIG = 'CheckoutConfig';
export const ROUTING_CONFIG_SERVICE = 'RoutingConfigService';
export const PAYMENT_METHOD_COMPONENT = 'PaymentMethodComponent';
export const USER_PAYMENT_SERVICE = 'UserPaymentService';
export const GLOBAL_MESSAGE_SERVICE = 'GlobalMessageService';
export const CHECKOUT_AUTH_GUARD = 'CheckoutAuthGuard';
export const CART_PAGE_LAYOUT_HANDLER = 'CartPageLayoutHandler';
export const SELECTIVE_CART_SERVICE = 'SelectiveCartService';
export const CDS_SPARTACUS_EVENT_SERVICE = 'SpartacusEventService';
export const CONSENT_SERVICE = 'ConsentService';
export const CDS_CONFIG = 'CdsConfig';
export const ADDRESS_BOOK_COMPONENT_SERVICE = 'AddressBookComponentService';
export const CHECKOUT_GUARD = 'CheckoutGuard';
export const EXPRESS_CHECKOUT_SERVICE = 'ExpressCheckoutService';
export const CONSENT_MANAGEMENT_FORM_COMPONENT =
  'ConsentManagementFormComponent';
/***** Classes end *****/

/***** Removed public api start *****/
export const CART_EFFECTS = 'CartEffects';
export const WISHLIST_EFFECTS = 'WishlistEffects';
export const CART_VOUCHER_EFFECTS = 'CartVoucherEffects';
export const CART_ENTRY_EFFECTS = 'CartEntryEffects';
export const CART_COMBINED_EFFECTS = 'effects';

export const GET_REDUCERS = 'getReducers';
export const REDUCER_TOKEN = 'reducerToken';
export const REDUCER_PROVIDER = 'reducerProvider';
export const CLEAR_CART_STATE = 'clearCartState';
export const META_REDUCERS = 'metaReducers';
export const CLEAR_MULTI_CART_STATE = 'clearMultiCartState';
export const MULTI_CART_META_REDUCERS = 'multiCartMetaReducers';
export const MULTI_CART_REDUCER_TOKEN = 'multiCartReducerToken';
export const GET_MULTI_CART_REDUCERS = 'getMultiCartReducers';
export const MULTI_CART_REDUCER_PROVIDER = 'multiCartReducerProvider';

export const SITE_CONTEXT_PARAMS_PROVIDERS = 'siteContextParamsProviders';
export const INIT_SITE_CONTEXT_ROUTES_HANDLER = 'initSiteContextRoutesHandler';
export const INITITIALIZE_CONTEXT = 'inititializeContext';
export const CONTEXT_SERVICE_PROVIDERS = 'contextServiceProviders';
/***** Removed public api end *****/

/***** Properties start *****/
export const CMS_COMPONENT_DATA_PROPERTY_NAME = 'componentData';
/***** Properties end *****/

/***** APIs start *****/
export const TODO_SPARTACUS = 'TODO:Spartacus -';

export const GET_COMPONENT_STATE_OLD_API = 'getComponentState';
export const GET_COMPONENTS_STATE_NEW_API = 'getComponentsState';

export const GET_COMPONENT_ENTITIES_OLD_API = 'getComponentEntities';
export const GET_COMPONENT_ENTITIES_COMMENT = `// ${TODO_SPARTACUS} '${GET_COMPONENT_ENTITIES_OLD_API}' has been removed, please use some of the newer API methods.`;

export const COMPONENT_STATE_SELECTOR_FACTORY_OLD_API =
  'componentStateSelectorFactory';
export const COMPONENTS_STATE_SELECTOR_FACTORY_NEW_API =
  'componentsLoaderStateSelectorFactory';

export const COMPONENT_SELECTOR_FACTORY_OLD_API = 'componentSelectorFactory';
export const COMPONENTS_SELECTOR_FACTORY_NEW_API = 'componentsSelectorFactory';
/***** APIs end *****/

/***** Misc start *****/
export const CSS_DOCS_URL =
  'https://sap.github.io/cloud-commerce-spartacus-storefront-docs/deprecation-guide/css';
/***** Misc end *****/
