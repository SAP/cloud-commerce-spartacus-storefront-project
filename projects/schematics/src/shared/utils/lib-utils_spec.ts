import { Tree } from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { Schema as SpartacusOptions } from '../../add-spartacus/schema';
import { UTF_8 } from '../constants';
import { SPARTACUS_FEATURES_MODULE } from '../feature-libs-constants';
import {
  addLibraryFeature,
  FeatureConfig,
  LibraryOptions,
  orderInstalledFeatures,
  shouldAddFeature,
} from './lib-utils';

const appModulePath = 'src/app/app.module.ts';
const spartacusFeaturesPath = `src/app/spartacus/${SPARTACUS_FEATURES_MODULE}.module.ts`;

describe('Lib utils', () => {
  const schematicRunner = new SchematicTestRunner(
    'schematics',
    path.join(__dirname, '../../collection.json')
  );

  let appTree: UnitTestTree;

  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const spartacusDefaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  const CLI_FEATURE_NAME = 'xxx-cli';
  const FEATURE_NAME = 'xxx';
  const FEATURE_FOLDER_NAME = 'xxx';
  const FEATURE_MODULE_NAME = 'XxxModule';
  const FEATURE_MODULE_IMPORT_PATH = '@spartacus/xxx';
  const ROOT_MODULE_NAME = 'XxxModuleRoot';
  const ROOT_FEATURE_MODULE_IMPORT_PATH = '@spartacus/xxx/root';
  const I18N_RESOURCES = 'translations';
  const I18N_CHUNKS = 'translationChunk';
  const ASSETS_IMPORT_PATH = '@spartacus/xxx/assets';
  const SCSS_FILE_NAME = 'xxx.scss';
  const STYLE_IMPORT_PATH = FEATURE_MODULE_IMPORT_PATH;

  const scssFilePath = `src/styles/spartacus/${SCSS_FILE_NAME}`;

  const BASE_FEATURE_CONFIG: FeatureConfig = {
    folderName: FEATURE_FOLDER_NAME,
    moduleName: FEATURE_NAME,
    featureModule: {
      name: FEATURE_MODULE_NAME,
      importPath: FEATURE_MODULE_IMPORT_PATH,
    },
    rootModule: {
      name: ROOT_MODULE_NAME,
      importPath: ROOT_FEATURE_MODULE_IMPORT_PATH,
    },
    i18n: {
      resources: I18N_RESOURCES,
      chunks: I18N_CHUNKS,
      importPath: ASSETS_IMPORT_PATH,
    },
    styles: {
      scssFileName: SCSS_FILE_NAME,
      importStyle: STYLE_IMPORT_PATH,
    },
  };

  const BASE_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    features: [CLI_FEATURE_NAME],
    lazy: true,
  };

  const DP_FEATURE_CONFIG: FeatureConfig = {
    folderName: 'dp',
    moduleName: 'DigitalPayments',
    featureModule: {
      name: 'DigitalPaymentsModule',
      importPath: '@spartacus/digital-payments',
    },
  };
  const DP_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    features: ['dp-cli'],
    lazy: true,
  };

  const CHECKOUT_FEATURE_CONFIG: FeatureConfig = {
    folderName: 'checkout',
    moduleName: 'Checkout',
    featureModule: {
      name: 'CheckoutModule',
      importPath: '@spartacus/Checkout/base',
    },
    rootModule: {
      name: 'CheckoutRootModule',
      importPath: '@spartacus/Checkout/base/root',
    },
  };
  const CHECKOUT_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    features: ['checkout-base-cli'],
    lazy: true,
  };

  const CART_FEATURE_CONFIG: FeatureConfig = {
    folderName: 'cart',
    moduleName: 'Cart',
    featureModule: {
      name: 'CartBaseModule',
      importPath: '@spartacus/cart/base',
    },
    rootModule: {
      name: 'CartBaseRootModule',
      importPath: '@spartacus/cart/base/root',
    },
  };
  const CART_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    features: ['cart-cli'],
    lazy: true,
  };

  const USER_PROFILE_FEATURE_CONFIG: FeatureConfig = {
    folderName: 'user',
    moduleName: 'UserProfile',
    featureModule: {
      name: 'UserProfileModule',
      importPath: '@spartacus/user',
    },
    rootModule: {
      name: 'UserProfileRootModule',
      importPath: '@spartacus/user/root',
    },
  };
  const USER_PROFILE_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    features: ['user-profile-cli'],
    lazy: true,
  };

  const ORDER_FEATURE_CONFIG: FeatureConfig = {
    folderName: 'order',
    moduleName: 'Order',
    featureModule: {
      name: 'OrderModule',
      importPath: '@spartacus/order',
    },
    rootModule: {
      name: 'OrderRootModule',
      importPath: '@spartacus/order/root',
    },
  };
  const ORDER_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    features: ['order-cli'],
    lazy: true,
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync(
        'add-spartacus',
        { ...spartacusDefaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();
  });

  describe('shouldAddFeature', () => {
    it('should return true if the feature is present in the given features array', () => {
      const feature1 = 'feature1';
      const features = [feature1];
      expect(shouldAddFeature(feature1, features)).toBe(true);
    });
    it('should return false if the feature is NOT present in the given features array', () => {
      const random = 'random';
      const feature1 = 'feature1';
      const features = [feature1];
      expect(shouldAddFeature(random, features)).toBe(false);
    });
  });

  describe('addLibraryFeature', () => {
    it('should add i18n config in feature module', async () => {
      const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
      const tree = await schematicRunner.callRule(rule, appTree).toPromise();

      // TODO: Finish when config util will be created
      const appModule = tree.read(appModulePath)?.toString(UTF_8);
      expect(appModule).not.toContain(
        `import { ${I18N_RESOURCES} } from '${ASSETS_IMPORT_PATH}';`
      );
    });
    it('should NOT add i18n if the config is not present', async () => {
      const featureConfig: FeatureConfig = {
        ...BASE_FEATURE_CONFIG,
        i18n: undefined,
      };
      const rule = addLibraryFeature(BASE_OPTIONS, featureConfig);
      const tree = await schematicRunner.callRule(rule, appTree).toPromise();

      // TODO: Finish when config util will be created
      const appModule = tree.read(appModulePath)?.toString(UTF_8);
      expect(appModule).not.toContain(`providers: [
        provideConfig({
          i18n: {`);
    });
    describe('when the lazy loading is configured', () => {
      it('should add it in the lazy loading way', async () => {
        const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
        const tree = await schematicRunner.callRule(rule, appTree).toPromise();

        const appModule = tree.read(appModulePath)?.toString(UTF_8);
        // TODO: Finish when config util will be created
        expect(appModule).not.toContain(
          `import { ${ROOT_MODULE_NAME} } from '${ROOT_FEATURE_MODULE_IMPORT_PATH}';`
        );
      });
    });
    describe('when the eager loading is configured', () => {
      it('should add it in the eager way', async () => {
        const rule = addLibraryFeature(
          { ...BASE_OPTIONS, lazy: false },
          BASE_FEATURE_CONFIG
        );
        const tree = await schematicRunner.callRule(rule, appTree).toPromise();

        const appModule = tree.read(appModulePath)?.toString(UTF_8);
        // TODO: Finish when config util will be created
        expect(appModule).not.toContain(
          `module: () => import('${FEATURE_MODULE_IMPORT_PATH}').then(`
        );
      });
    });
    describe('style', () => {
      describe('when style config is provided', () => {
        describe('and the scss file does NOT exist', () => {
          it('should add it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            const tree = await schematicRunner
              .callRule(rule, appTree)
              .toPromise();

            expect(tree.exists(scssFilePath)).toEqual(true);
            const content = tree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(`@import "${FEATURE_MODULE_IMPORT_PATH}";`);
          });
        });
        describe('and the scss with the same content already exists', () => {
          beforeEach(() => {
            appTree.create(
              scssFilePath,
              `@import "${FEATURE_MODULE_IMPORT_PATH}";`
            );
          });
          it('should NOT append it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            const tree = await schematicRunner
              .callRule(rule, appTree)
              .toPromise();

            expect(tree.exists(scssFilePath)).toEqual(true);
            const content = tree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(`@import "${FEATURE_MODULE_IMPORT_PATH}";`);
          });
        });
        describe('and the scss file with a different content already exists', () => {
          const randomContent = `@import "@random/xxx";`;
          beforeEach(() => {
            appTree.create(scssFilePath, randomContent);
          });
          it('should append it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            const tree = await schematicRunner
              .callRule(rule, appTree)
              .toPromise();

            expect(tree.exists(scssFilePath)).toEqual(true);
            const content = tree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(
              `${randomContent}\n@import "${FEATURE_MODULE_IMPORT_PATH}";`
            );
          });
        });
      });
      describe('when style config is NOT provided', () => {
        it('should not add it', async () => {
          const rule = addLibraryFeature(BASE_OPTIONS, {
            ...BASE_FEATURE_CONFIG,
            styles: undefined,
          });
          const tree = await schematicRunner
            .callRule(rule, appTree)
            .toPromise();

          expect(tree.exists(scssFilePath)).toEqual(false);
        });
      });
    });
  });

  describe('feature ordering', () => {
    let tree: Tree;
    beforeEach(async () => {
      tree = await schematicRunner
        .callRule(addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG), appTree)
        .toPromise();
      tree = await schematicRunner
        .callRule(addLibraryFeature(DP_OPTIONS, DP_FEATURE_CONFIG), tree)
        .toPromise();
      tree = await schematicRunner
        .callRule(
          addLibraryFeature(CHECKOUT_OPTIONS, CHECKOUT_FEATURE_CONFIG),
          tree
        )
        .toPromise();
      tree = await schematicRunner
        .callRule(addLibraryFeature(CART_OPTIONS, CART_FEATURE_CONFIG), tree)
        .toPromise();
      tree = await schematicRunner
        .callRule(
          addLibraryFeature(USER_PROFILE_OPTIONS, USER_PROFILE_FEATURE_CONFIG),
          tree
        )
        .toPromise();
      tree = await schematicRunner
        .callRule(addLibraryFeature(ORDER_OPTIONS, ORDER_FEATURE_CONFIG), tree)
        .toPromise();

      tree = await schematicRunner
        .callRule(orderInstalledFeatures(spartacusDefaultOptions), tree)
        .toPromise();
    });

    it('should appropriately order the feature modules', () => {
      expect(
        tree.read(spartacusFeaturesPath)?.toString(UTF_8)
      ).toMatchSnapshot();
    });
  });
});
