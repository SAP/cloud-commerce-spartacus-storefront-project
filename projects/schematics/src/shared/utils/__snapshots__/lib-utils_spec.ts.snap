// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lib utils addLibraryFeature custom config option should add the custom config when set 1`] = `
"import { NgModule } from '@angular/core';
import { CdsConfig } from \\"@spartacus/cds\\";
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.XxxModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  }),
  provideConfig(<CdsConfig>{
    cds: {
      profileTag: {
        javascriptUrl:
          'random',
        configUrl:
          'random',
        allowInsecureCookies: true,
      },
    },
  })
  ]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature recreate option should remove the feature module and recreate it 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.XxxModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  })
  ]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature recreate option should remove the feature module and recreate it 2`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.YyyModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  })
  ]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature should NOT add i18n if the config is not present 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, provideConfig } from \\"@spartacus/core\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.XxxModule),
      },
    }
  })]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature when the eager loading is configured should add it in the eager way 1`] = `
"import { NgModule } from '@angular/core';
import { I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { XxxModule } from \\"@spartacus/xxx\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot,
    XxxModule
  ],
  providers: [provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  })]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature when the lazy loading is configured should add it in the lazy loading way 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.XxxModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  })
  ]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils feature ordering should appropriately order the feature modules 1`] = `
"import { NgModule } from '@angular/core';
import { AnonymousConsentsModule, AuthModule, CostCenterOccModule, ExternalRoutesModule, ProductModule, ProductOccModule, UserModule, UserOccModule } from \\"@spartacus/core\\";
import { AddressBookModule, AnonymousConsentManagementBannerModule, AnonymousConsentsDialogModule, BannerCarouselModule, BannerModule, BreadcrumbModule, CategoryNavigationModule, CmsParagraphModule, ConsentManagementModule, FooterNavigationModule, HamburgerMenuModule, HomePageEventModule, LinkModule, LoginRouteModule, LogoutModule, MyCouponsModule, MyInterestsModule, NavigationEventModule, NavigationModule, NotificationPreferenceModule, PaymentMethodsModule, ProductCarouselModule, ProductDetailsPageModule, ProductFacetNavigationModule, ProductImagesModule, ProductIntroModule, ProductListingPageModule, ProductListModule, ProductPageEventModule, ProductReferencesModule, ProductSummaryModule, ProductTabsModule, SearchBoxModule, SiteContextSelectorModule, StockNotificationModule, TabParagraphContainerModule } from \\"@spartacus/storefront\\";
import { CartFeatureModule } from './features/cart/cart-feature.module';
import { CheckoutFeatureModule } from './features/checkout/checkout-feature.module';
import { DigitalPaymentsFeatureModule } from './features/dp/digital-payments-feature.module';
import { OrderFeatureModule } from './features/order/order-feature.module';
import { UserProfileFeatureModule } from './features/user/user-profile-feature.module';
import { XxxFeatureModule } from './features/xxx/xxx-feature.module';

@NgModule({
  declarations: [],
  imports: [AuthModule.forRoot(),
    LogoutModule,
    LoginRouteModule,
    HamburgerMenuModule,
    SiteContextSelectorModule,
    LinkModule,
    BannerModule,
    CmsParagraphModule,
    TabParagraphContainerModule,
    BannerCarouselModule,
    CategoryNavigationModule,
    NavigationModule,
    FooterNavigationModule,
    BreadcrumbModule,
    UserModule,
    UserOccModule,
    AddressBookModule,
    PaymentMethodsModule,
    NotificationPreferenceModule,
    MyInterestsModule,
    StockNotificationModule,
    ConsentManagementModule,
    MyCouponsModule,
  AnonymousConsentsModule.forRoot(),
    AnonymousConsentsDialogModule,
    AnonymousConsentManagementBannerModule,
  ProductModule.forRoot(),
    ProductOccModule,
    ProductDetailsPageModule,
    ProductListingPageModule,
    ProductListModule,
    SearchBoxModule,
    ProductFacetNavigationModule,
    ProductTabsModule,
    ProductCarouselModule,
    ProductReferencesModule,
    ProductImagesModule,
    ProductSummaryModule,
    ProductIntroModule,
    CostCenterOccModule,
    NavigationEventModule,
    HomePageEventModule,
    ProductPageEventModule,
  ExternalRoutesModule.forRoot(),
    UserProfileFeatureModule,
    CartFeatureModule,
    OrderFeatureModule,
    CheckoutFeatureModule,
    DigitalPaymentsFeatureModule,
    XxxFeatureModule]
})
export class SpartacusFeaturesModule { }
"
`;
